cryptography/src/bls12381/primitives/ops.rs

```diff
@@ -18,77 +25,163 @@ pub fn keypair<R: RngCore>(rng: &mut R) -> (group::Private, group::Public) {
+pub fn sign(private: &group::Private, dst: DST, payload: &[u8]) -> group::Signature {
+    let mut s = group::Signature::zero();
+    s.map(dst, payload);
+    s.mul(private);
+    s
+}
+
+/// Verify the signature from the provided public key.
+pub fn verify(
+    public: &group::Public,
+    dst: DST,
+    payload: &[u8],
+    signature: &group::Signature,
+) -> Result<(), Error> {
+    let mut hm = group::Signature::zero();
+    hm.map(dst, payload);
+    if !equal(public, signature, &hm) {
+        return Err(Error::InvalidSignature);
+    }
+    Ok(())
+}
+
+/// Generates a proof of possession for the private key.
+pub fn sign_proof_of_possession(private: &group::Private) -> group::Signature {
+    // Get public key
+    let mut public = group::Public::one();
+    public.mul(private);
+
+    // Sign the public key
+    sign(private, PROOF_OF_POSSESSION, public.serialize().as_slice())
+}
+
+/// Verifies a proof of possession for the provided public key.
+pub fn verify_proof_of_possession(
+    public: &group::Public,
+    signature: &group::Signature,
+) -> Result<(), Error> {
+    verify(
+        public,
+        PROOF_OF_POSSESSION,
+        public.serialize().as_slice(),
+        signature,
+    )
+}
+
 /// Signs the provided message with the private key.
 ///
-/// The message is hashed according to RFC 9380.
-///
 /// # Determinism
 ///
 /// Signatures produced by this function are deterministic and are safe
 /// to use in a consensus-critical context.
-pub fn sign(
+pub fn sign_message(
     private: &group::Private,
     namespace: Option<&[u8]>,
     message: &[u8],
 ) -> group::Signature {
-    let mut s = group::Signature::zero();
-    match namespace {
-        Some(namespace) => s.map(&union_unique(namespace, message)),
-        None => s.map(message),
+    let payload = match namespace {
+        Some(namespace) => Cow::Owned(union_unique(namespace, message)),
+        None => Cow::Borrowed(message),
     };
-    s.mul(private);
-    s
+    sign(private, MESSAGE, &payload)
 }
 
 /// Verifies the signature with the provided public key.
-pub fn verify(
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on
+/// `public` and `signature`.
+pub fn verify_message(
     public: &group::Public,
     namespace: Option<&[u8]>,
     message: &[u8],
     signature: &group::Signature,
 ) -> Result<(), Error> {
-    let mut hm = group::Signature::zero();
-    match namespace {
-        Some(namespace) => hm.map(&union_unique(namespace, message)),
-        None => hm.map(message),
+    let payload = match namespace {
+        Some(namespace) => Cow::Owned(union_unique(namespace, message)),
+        None => Cow::Borrowed(message),
     };
-    if !equal(public, signature, &hm) {
-        return Err(Error::InvalidSignature);
+    verify(public, MESSAGE, &payload, signature)
+}
+
+/// Generates a proof of possession for the private key share.
+pub fn partial_sign_proof_of_possession(
```

Comment:
We support generating/verifying threshold proof of possessions because this is required for some party to utilize a threshold signature in some function that employs multi-signatures.